Adding `tbl_listing()`

ddsjoberg opened the pull request:
**What changes are proposed in this pull request?**
* Adding function `tbl_listing()` to report data listings from a data frame.

**If there is an GitHub issue associated with this pull request, please provide link.**
closes #124

```r
library(dplyr, warn.conflicts = FALSE)

tbl_listing_ex1 <-
  head(df_adverse_events, n = 10) %>%
  select(system_organ_class, adverse_event, grade, drug_attribution, patient_id) %>%
  arrange(adverse_event, desc(grade)) %>%
  tbl_listing(group_by = system_organ_class) %>%
  bold_labels()
```
![image](https://user-images.githubusercontent.com/26774684/164336319-d36dc7ec-fc9d-4ff1-a51b-16e07c1b0645.png)


I got a little fancy with this example. Also, update however you'd like :)
```r
set.seed(11234)
tbl_listing_ex2 <-
df_patient_characteristics %>%
  dplyr::slice_sample(n = 10) %>%
  select(patient_id, status, discontinued, off_trt_ae) %>%
  tbl_listing() %>%
  as_gt() %>%
  gt::opt_row_striping()
```
![image](https://user-images.githubusercontent.com/26774684/164336403-fe50a707-45e9-48a4-aea1-5fe7c79e4e56.png)


--------------------------------------------------------------------------------

Checklist for PR reviewer

- [ ] Ensure all package dependencies are installed by running `renv::install()`
- [ ] PR branch has pulled the most recent updates from main branch. Ensure the pull request branch and your local version match and both have the latest updates from the main branch.
- [ ] If a new function was added, function included in `_pkgdown.yml`
- [ ] If a bug was fixed, a unit test was added for the bug check
- [ ] Run `pkgdown::build_site()`. Check the R console for errors, and review the rendered website.
- [ ] Code coverage is suitable for any new functions/features. Review coverage with `covr::report()`. Before you run, set `Sys.setenv(NOT_CRAN="true")` and begin in a fresh R session without any packages loaded. 
- [ ] R CMD Check runs without errors, warnings, and notes
- [ ] `usethis::use_spell_check()` runs with no spelling errors in documentation
- [ ] Has `NEWS.md` been updated with the changes from this pull request under the heading "`# gtreg (development version)`". If there is an issue associated with the pull request, reference it in parentheses at the end update (see `NEWS.md` for examples).
- [ ] Has the version number been incremented using `usethis::use_version(which = "dev")` 
- [ ] Approve Pull Request
- [ ] Merge the PR. Please use "Squash and merge".

ddsjoberg commented:
![MicrosoftTeams-image](https://user-images.githubusercontent.com/26774684/164303335-8d56470e-5a95-491d-bd0a-c8880f0f4734.png)

shannonpileggi commented:
i do not know why but i am so silly excited about this function!!!

i think we chatted about bold labels, but I don't remember the conversation exactly.

the function example shows used of piped `bold_labels()`

```
tbl_listing_ex1 <-
  head(df_adverse_events, n = 10) %>%
  select(system_organ_class, adverse_event, grade, drug_attribution, patient_id) %>%
  arrange(adverse_event, desc(grade)) %>%
  tbl_listing(group_by = system_organ_class) %>%
  bold_labels()
```

but the function documentation has the `bold_headers = TRUE` argument.

Which way do we want to proceed with bolding - via piped call or function argument? I can make updates accordingly if you want 😊

ddsjoberg commented:
I love it too!

```r
@param bold_headers logical indicating whether to bold column headers.
```
That argument controls bolding the column headers and the `bold_labels()` function bolds the label rows (the grouped data).

shannonpileggi commented:
omg we DID talk about this! more ☕☕☕ please, and forgive me!

shannonpileggi commented:
Regarding the documentation "Both the grouping variable and the first column must be character." 

Would it be possible to submit the grouping variable as a factor to control the ordering of the grouping variable? Because otherwise the ordering of the grouping variable will always be alphabetical, correct?

Currently this errors out as expected b/c soc is a factor, but otherwise this is the intent.

```
head(df_adverse_events, n = 10) %>%
  select(system_organ_class, adverse_event, grade, drug_attribution, patient_id) %>%
  mutate(system_organ_class = fct_relevel(system_organ_class, "Gastrointestinal disorders")) %>% 
  tbl_listing(group_by = system_organ_class, bold_headers = TRUE)
```

ddsjoberg commented:
Great point! I removed that condition and added a note to the documentation file that the first column and the grouping variable column will be combined and the class/type may be updated.

shannonpileggi commented:
It looks like character/factor are coercing well! 🎉

```
head(df_adverse_events, n = 10) %>%
  select(system_organ_class, adverse_event, grade, drug_attribution, patient_id) %>%
  mutate(system_organ_class = fct_relevel(system_organ_class, "Gastrointestinal disorders")) %>% 
  tbl_listing(group_by = system_organ_class, bold_headers = TRUE)
```

works great w/o error.

When the group_by and first column are numeric this also works without errror.

```
head(df_adverse_events, n = 10) %>%
  mutate(num = rep(1:2, 5)) %>% 
  select(num, grade, system_organ_class, adverse_event, drug_attribution, patient_id) %>%
  tbl_listing(group_by = num, bold_headers = TRUE)
```

but if the user mixes and matches numeric/character for group_by and first column, there is an error.

```
head(df_adverse_events, n = 10) %>%
  mutate(num = rep(1:2, 5)) %>% 
  select(num, system_organ_class, adverse_event, grade, drug_attribution, patient_id) %>%
  tbl_listing(group_by = num, bold_headers = TRUE)
```

Error in `dplyr::bind_rows()`:
! Can't combine `..1$system_organ_class` <integer> and `..2$system_organ_class` <character>.
Run `rlang::last_error()` to see where the error occurred.

```
head(df_adverse_events, n = 10) %>%
  mutate(num = rep(1:2, 5)) %>% 
  select(system_organ_class, num, system_organ_class, adverse_event, grade, drug_attribution, patient_id) %>%
  tbl_listing(group_by = system_organ_class, bold_headers = TRUE)
```

Error in `dplyr::bind_rows()`:
! Can't combine `..1$num` <character> and `..2$num` <integer>.
Run `rlang::last_error()` to see where the error occurred.

Is this something we want to try to resolve or exclude this option and provide messaging?

ddsjoberg commented:
hey hey hey @shannonpileggi !

i made an update that allows for the combining any two column types and added your examples above as unit tests.

TBH, the line of code is horrific to read...but it's just one line! Instead of using the typical dplyr verbs for stacking two data frames, i opted for `rbind()` (which allows any type to be stacked). But to use `rbind()`, I had to add all the missing columns to the header data frame. Anyway, it works. If you know a better way, please let me know!

shannonpileggi commented on R/tbl_listing.R:
```diff
@@ -0,0 +1,129 @@
+#' column in the table will be combined and the type/class may be converted
+#' to common type/class for both columns.
+#'
+#' Both the grouping variable and the
+#' first column must be character.
```
Is this a fair description?

```suggestion
#' When the `group_by` argument is utilized, groups are ordered accrording to the grouping 
#'  variable's type (i.e.,  character, numeric, or factor).
```

shannonpileggi commented on R/tbl_listing.R:
```diff
@@ -0,0 +1,129 @@
+#' Data Listing Table
+#'
+#' Function creates a gtsummary-class listing of data. Column labels are
+#' used as column headers, when present.
```
```suggestion
#' used as column headers, when present. The listing prints observations in the order
of the input data.
```

shannonpileggi reviewed (approved):
I only have a couple of documentation suggestions - feel free to tweak or omit, not sure I'm being the most articulate right now.

Otherwise, this looks great!! And sorry, I don't have any suggestions for the `rbind` beast at this time, but at least it works!
